### ISSUE TITLE
Be able to create an element from an element template

### ISSUE THREAD

**Issue by nikku** [bpmn-io/bpmn-js-properties-panel#582]
__Is your feature request related to a problem? Please describe__

As a user I am able to change an element to an element template of a specific type. I am also able to declare an element template _as default_; this ensures it is applied as the element is created.

What I am not able to do yet is to create an element of a specific type, i.e. because I want to create it right off the palette or append it:

![capture Vzsshq_optimized](https://user-images.githubusercontent.com/58601/153565135-febf3d23-56d1-4802-b87f-1c72bbe75abb.gif)

We have a workaround for what I envision: Bootstrap the element (with all required extensions) manually, cf. [`ConnectorsExtension`](https://github.com/bpmn-io/connectors-modeling-demo/blob/master/connectors-extension/ConnectorsExtension.js#L196). That, however forces me to know all element templates in advance; it is not a solution applicable for general template support.

__Describe the solution you'd like__

* [ ] It is possible to create an element of a specific type similar to `ElementFactory#create**`, i.e. `elementTemplates.createElement(template: ElementTemplate): DiagramElement` - The element shall not be added to the canvas yet, but it shall be already configured to the extend needed. 

__Describe alternatives you've considered__

Decide this is not a direction we want to go.

__Additional context__

This pops up in the context of [connectors](https://github.com/orgs/camunda/projects/14/views/4), specifically in the realm of [how to integrate them nicely into the existing modeling UX](https://github.com/camunda/web-modeler/issues/364#issuecomment-1036017796).

**Comment by MaxTru** [comment]
Next step: discuss priority / necessity of this

**Comment by pinussilvestrus** [comment]
Just for my understanding, can you maybe explain what would be the difference between our existing commands?

```js
const element = elementFactory.createShape({ type: "bpmn:Task" });

return commandStack.execute('propertiesPanel.camunda.changeTemplate', {
  element: element,
  newTemplate: template
});
```

The `ChangeElementTemplateHandler` would be then create all the necessary properties retrieved from the bindings.  The difference here is probably that I need to create the (base) element beforehand?

**Comment by nikku** [comment]
To clarify: This feature request is about creating a diagram element _of a specific type_ outside the core modeling cycle.

The use-case is that I want to pull that element out from the palette (into "hover state"), and preview it accordingly (or add it using existing modeling APIs).

**Comment by pinussilvestrus** [comment]
Thanks for the clarification :+1:

I quickly sketched this one from my current understanding

```js
/**
 * Create an element based on an element template.
 *
 * @param {ElementTemplate} template
 * @returns {djs.model.Base}
 */
createElement(template) {

  // todo: throw?
  if (!template) {
    return;
  }

  const {
    appliesTo
  } = template;

  // todo: throw?
  if (!appliesTo || !appliesTo.length) {
    return;
  }

  // todo: what do if appliesTo has more then one entries?
  const type = appliesTo[0];

  // (1) create element from appliesTo
  const element = this._elementFactory.createShape({ type });

  // (2) apply template
  // todo: make command configurable, e.g. propertiesPanel.zeebe.changeTemplate
  this._commandStack.execute('propertiesPanel.camunda.changeTemplate', {
    element: element,
    newTemplate: template
  });

  return element;
}
```

This already works pretty nicely, e.g. by using [this example template](https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/test/spec/provider/element-templates/fixtures/field-injections.json)

![image](https://user-images.githubusercontent.com/9433996/155728691-970f3faa-5606-437f-a9c7-3c9c91725a99.png)

**Comment by nikku** [comment]
But do we want to execute a commandStack action to get the job done?

We don't usually do that before element creation? What is supposed to be undone here?

**Comment by pinussilvestrus** [comment]
Good point. That would be the easiest way though, otherwise, we would need to pull [all the stuff](https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/src/provider/element-templates/cmd/ChangeElementTemplateHandler.js#L75-L103) out of the command, and make it generally available. That would be another option.

**Comment by pinussilvestrus** [comment]
If we want to make this stuff available without any command being involved, we would need to build from scratch (again), for all bindings.

**Comment by pinussilvestrus** [comment]
I tried out an approach that _doesn't_ use commands, but re-uses our existing `CreateHelper` to fill up the element by the defined bindings.

**tldr**
```js
const elementTemplates = modeler.get('elementTemplates');
const create = modeler.get('create');

const element = elementTemplates.createElement(template);
create.start(canvasEvent({ x: 250, y: 300 }), element);
```

![Kapture 2022-03-03 at 14 15 24](https://user-images.githubusercontent.com/9433996/156572084-42054777-990b-434f-ad2b-3903f0a13b21.gif)

Find some sketching around on this [Miro Board](https://miro.com/app/board/uXjVOIOoGto=/).
Branch: https://github.com/bpmn-io/bpmn-js-properties-panel/compare/582-element-templates-create?expand=1
What's missing:
* all bindings for Camunda Platform
* optional for `zeebe:input` & `zeebe:output`
* shall we validate the template beforehand?
* do not create extension elements when not needed (keep XML clean)

Let's sync on what direction we should follow next week @nikku 👍

**Comment by pinussilvestrus** [comment]
After a chat with @nikku I think the direction described in https://github.com/bpmn-io/bpmn-js-properties-panel/issues/582#issuecomment-1058035853 is the way to go. Furthermore, we could consider (as future improvement)
* refactor the existing `propertiesPanel.x.changeTemplate` command to a similar fashion, moving the binding related bits to providers
* promote the `elementTemplates` feature as a service to a higher extent, e.g. providing more API in a similar fashion to the `createElement` one
  * `elementTemplates.getMatching(element): Array<ElementTemplate> // get all applicable element templates`
  * `elementTemplates.apply(element, template): void // apply my template`
  * ...

> all bindings for Camunda Platform

Since providing this in a similar fashion as for `cloud-element-templates` would need more time (due to some more bindings + edge cases), we could consider going [the "command" solution](https://github.com/bpmn-io/bpmn-js-properties-panel/issues/582#issuecomment-1050882669), **but providing the same API**. We can easily refactor it internally to the factory approach once we have more time. I would have to time box to make a final decision.

**Comment by pinussilvestrus** [comment]
Along with our Reflect & Plan, and given our current priorities, I will descope the API for Camunda Platform for now. I added another issue for that: https://github.com/bpmn-io/bpmn-js-properties-panel/issues/607
